src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c

```diff
@@ -630,13 +639,58 @@ static void rr_connectivity_changed_locked(grpc_exec_ctx *exec_ctx, void *arg,
         GRPC_CLOSURE_SCHED(exec_ctx, pp->on_complete, GRPC_ERROR_NONE);
         gpr_free(pp);
       }
+      rr_subchannel_list_unref(exec_ctx, sd->subchannel_list,
+                               "sd_shutdown+started_picking");
+      // unref the "rr_connectivity_update" weak ref from start_picking.
+      GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &p->base,
+                                "rr_connectivity_sd_shutdown");
+    } else {
+      // Policy isn't shutting down, we aren't in connectivity shutdown (from
```

Comment:
The moment `num_shutdown` equals `num_subchannels`, the policy will go into shutdown. Given that initially `num_shutdown < num_subchannels`, and that `num_shutdown` is only incremented by one _before_ performing the `num_shutdown == num_subchannels` test that'd put the policy in shutdown, it's impossible for `num_shutdown` to ever be `≥ num_subchannels`. 

If we have a single subchannel and it reports as shutdown, that event will be captured by ` rr_connectivity_changed_locked`, which will first call `update_state_counters_locked`, making `num_shutdown` be 1. After that, `update_lb_connectivity_status_locked` will detect that `num_subchannels == num_shutdown`, which will make `p->shutdown` true and the subchannel will _not_ try to reconnect. In fact, this scenario is precisely what the `RoundRobinReresolve` (renamed it) testcase covers. Changing `kNumServers` to 1 would literally make it so, but any other value has the same effect (while adding more complexity to the dance behind the scenes). At the end of the day, `rr_connectivity_changed_locked` will be called with an initial `num_shutdown == num_subchannels - 1`, which will become an equality in the manner described above. This is independent of there being one or many subchannels in flux.

<hr>

All that being said, we may want to use something other than `num_subchannels` to compare against `num_shutdown` in order to put the policy in shutdown and trigger a re-resolution. For example, we may want to shutdown-and-reresolve after `num_subchannels/2` subchannels have gone into shutdown (note that this is orthogonal to reconnections, the same way subchannels currently bail out the moment they detect the policy is shutting down). This all is TBD and suitable for a separate PR.